File: ungrouped/3076.py
class Solution:
    m = {}
    def shortestSubstrings(self, arr: List[str]) -> List[str]:
        self.m = {}
        sub = []
        for s in arr:
            sub.append(self.cut(s))

        ans = [""] * len(arr)
        for i in range(len(sub)):
            for j in range(len(sub[i])):
                if self.m[sub[i][j]] == 1 and len(ans[i]) == 0:
                    ans[i] = sub[i][j]
                elif self.m[sub[i][j]] == 1 and len(ans[i]) > len(sub[i][j]):
                    ans[i] = sub[i][j]
                    
        return ans
    
    def cut(self, s: str):
        results = []
        flag = set()
        for x in range(len(s)):
            for i in range(len(s) - x):
                tmp = s[i:i + x + 1]
                results.append(tmp)
                if tmp not in self.m:
                    self.m[tmp] = 0
                if tmp not in flag:
                    self.m[tmp] += 1
                    flag.add(tmp)
                
        results.sort()
        return results

File: ungrouped/3076.cpp
#include "include.h"

// endlesscheng
class Solution {
public:
    vector<string> shortestSubstrings(vector<string> &arr) {
        int n = arr.size();
        auto check = [&](int i, string &sub) {
            for (int j = 0; j < n; j++) {
                if (j != i && arr[j].find(sub) != string::npos) {
                    return false;
                }
            }
            return true;
        };

        vector<string> ans(n);
        for (int i = 0; i < n; i++) {
            int m = arr[i].size();
            string res;
            for (int size = 1; size <= m && res.empty(); size++) {
                for (int j = size; j <= m; j++) {
                    string t = arr[i].substr(j - size, size);
                    if ((res.empty() || t < res) && check(i, t)) {
                        res = t;
                    }
                }
            }
            ans[i] = res;
        }
        return ans;
    }
};